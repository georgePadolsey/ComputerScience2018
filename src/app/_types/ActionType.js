// @flow
export type actionType = {
  +type: string
};
