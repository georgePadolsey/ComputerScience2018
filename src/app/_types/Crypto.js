// @flow

export type OHLCVCandle = [
  number, // UTC timestamp in milliseconds, integer
  number, // (O)pen price, float
  number, // (H)ighest price, float
  number, // (L)owest price, float
  number, // (C)losing price, float
  number // (V)olume (in terms of the base currency), float
];

export type TimedCandleData = {
  data: OHLCVCandle[],
  lastUpdated: number,
  timeframe: string
};

export type CryptoState = { +loadedExchanges: string[] };
