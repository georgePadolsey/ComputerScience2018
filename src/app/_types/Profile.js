// @flow

/**
 * # Types
 */

export type Profile = {
  displayName: string,
  uuid: string
};

export type ProfileData = {
  +currentProfile: ?string,
  +loadedProfiles: { [string]: Profile }
};
