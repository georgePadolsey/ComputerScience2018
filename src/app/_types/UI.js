// @flow
// import { PROFILE_CREATOR_STAGES } from '../actions/ui';

const PROFILE_CREATOR_STAGES = {
  ACCOUNT_ADDER: 'ACCOUNT_ADDER',
  ADD_BALANCE: 'ADD_BALANCE',
  ADD_EXCHANGE: 'ADD_EXCHANGE',
  ADD_WALLET: 'ADD_WALLET'
};

export type ProfileCreatorStage = $Keys<typeof PROFILE_CREATOR_STAGES>;

export type MainLayout = {
  [string]: Array<{
    w: number,
    h: number,
    x: number,
    y: number,
    i: string,
    moved: boolean,
    static: boolean
  }>
};

export type UIData = {
  +mainPanelEditMode: boolean,
  +mainPanelLayouts: MainLayout,
  +showAddMainChart: boolean,
  +offeredCreator: boolean,
  +showProfileCreator: boolean,
  +profileCreatorStage: ProfileCreatorStage,
  +firstTime: boolean
};
