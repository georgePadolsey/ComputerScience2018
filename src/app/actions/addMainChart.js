// @flow
import * as ACTIONS from './types/addMainChart';

type AddMainChartAction = {
  type: $Keys<typeof ACTIONS>
};
export function show(): AddMainChartAction {
  return setShow(true);
}

export function hide(): AddMainChartAction {
  return setShow(false);
}

export function setShow(showMainChart: boolean): AddMainChartAction {
  return {
    type: ACTIONS.ADD_MAIN_CHART_SET_SHOW,
    payload: showMainChart
  };
}

export function setSelectedExchange(selectedExchange: string): AddMainChartAction {
  return {
    type: ACTIONS.ADD_MAIN_CHART_SET_SELECTED_EXCHANGE,
    payload: selectedExchange
  };
}

export function setSelectedSymbol(selectedSymbol: string): AddMainChartAction {
  return {
    type: ACTIONS.ADD_MAIN_CHART_SET_SELECTED_SYMBOL,
    payload: selectedSymbol
  };
}
