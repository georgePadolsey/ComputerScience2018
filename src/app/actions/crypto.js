// @flow
import type { TimedCandleData } from '../_types/Crypto';

export const CACHE_OHLCV_DATA = 'CACHE_OHLCV_DATA';

export function cacheOHLCVData(
  exchangeId: string,
  symbol: string,
  timedCandleData: TimedCandleData[]
) {
  return {
    type: CACHE_OHLCV_DATA,
    payload: {
      exchangeId,
      symbol,
      timedCandleData
    }
  };
}
