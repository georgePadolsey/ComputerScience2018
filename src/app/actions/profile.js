// @flow
import getProfileData from '../utils/ProfileProvider';
import * as PROFILE_ACTIONS from './types/profile';
import type { Dispatch } from 'redux';
import type { ProfileData } from '../_types/Profile';

type ProfileAction = {
  +type: $Keys<typeof PROFILE_ACTIONS>
};

export function changeProfile(uuid: string): ProfileAction {
  return {
    type: PROFILE_ACTIONS.CHANGE_PROFILE,
    payload: uuid
  };
}

export function changeProfileName(
  uuid: string,
  displayName: string
): ProfileAction {
  return {
    type: PROFILE_ACTIONS.CHANGE_PROFILE_NAME,
    payload: {
      uuid,
      displayName
    }
  };
}

export function loadProfileData(): (Dispatch<*>) => Promise<void> {
  return async dispatch =>
    dispatch((loadedProfileData(await getProfileData()): Object));
}

export function loadedProfileData(profileData: ProfileData): ProfileAction {
  return {
    type: PROFILE_ACTIONS.LOADED_PROFILE_DATA,
    payload: profileData
  };
}

export function correctProfileData(): ProfileAction {
  return {
    type: PROFILE_ACTIONS.CORRECT_PROFILE_DATA
  };
}
