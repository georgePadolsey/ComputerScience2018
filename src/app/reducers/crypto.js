// @flow
import merge from 'lodash/merge';
import type { actionType } from '../_types/ActionType';
import type { CryptoState } from '../_types/Crypto';
import { LOADED_EXCHANGE } from '../actions/types/crypto';

const defaultCryptoState: CryptoState = {
  loadedExchanges: []
};

/**
 * This reducer is not saved to a data file
 * @param {*} state
 * @param {*} action
 */
export default function cryptoReducer(state: CryptoState = defaultCryptoState, action: actionType) {
  switch (action.type) {
    case LOADED_EXCHANGE:
      return merge({}, state, {
        loadedExchanges: [...state.loadedExchanges, ...action.payload]
      });
    default:
      return state;
  }
}
