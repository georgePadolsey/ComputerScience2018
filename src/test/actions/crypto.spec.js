// @flow
import * as actions from '../../app/actions/crypto';
import { Resolutions } from '../../app/_types/Crypto';

describe('Crypto actions', () => {
  it('should be able to make cache OHLCV data action', () => {
    expect(actions.cacheOHLCVData('fakeId', 'USD/BTC', [
      {
        data: [[5, 5, 5, 5, 5, 5]],
        from: 0,
        to: 5,
        expires: 6,
        resolution: Resolutions.HOUR.id()
      },
      {
        data: [[6, 6, 6, 6, 6, 6]],
        from: 0,
        to: 5,
        expires: 6,
        resolution: Resolutions.DAY.id()
      }
    ])).toMatchSnapshot();
  });
});
