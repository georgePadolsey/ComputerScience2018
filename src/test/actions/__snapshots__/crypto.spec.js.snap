// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crypto actions should be able to make cache OHLCV data action 1`] = `
Object {
  "payload": Object {
    "exchangeId": "fakeId",
    "symbol": "USD/BTC",
    "timedCandleData": Array [
      Object {
        "data": Array [
          Array [
            5,
            5,
            5,
            5,
            5,
            5,
          ],
        ],
        "expires": 6,
        "from": 0,
        "resolution": "HOUR",
        "to": 5,
      },
      Object {
        "data": Array [
          Array [
            6,
            6,
            6,
            6,
            6,
            6,
          ],
        ],
        "expires": 6,
        "from": 0,
        "resolution": "DAY",
        "to": 5,
      },
    ],
  },
  "type": "CACHE_OHLCV_DATA",
}
`;
